Sources/Sentry/SentrySession.m

```diff
@@ -9,61 +9,92 @@ @implementation SentrySession
-        }
-        NSString *status = [jsonObject valueForKey:@"status"];
-        if ([@"ok" isEqualToString:status]) {
-            _status = kSentrySessionStatusOk;
-        } else if ([@"exited" isEqualToString:status]) {
-            _status = kSentrySessionStatusExited;
-        } else if ([@"crashed" isEqualToString:status]) {
-            _status = kSentrySessionStatusCrashed;
-        } else if ([@"abnormal" isEqualToString:status]) {
-            _status = kSentrySessionStatusAbnormal;
-        }
-        _sequence = [[jsonObject valueForKey:@"seq"] unsignedIntegerValue];
-        _errors = [[jsonObject valueForKey:@"errors"] unsignedIntegerValue];
+    // We use the default constructor here to set the non nullable values to a default values,
+    // because this could cause crashes, for example, in serialize.
+    // With this approach we avoid crashes and accept the tradeoff that some session data might not
+    // be 100% accurate.
+    if (self = [self init]) {
+        NSUUID *sessionId = [[NSUUID UUID] initWithUUIDString:[jsonObject valueForKey:@"sid"]];
+        if (nil != sessionId) {
+            _sessionId = [[NSUUID UUID] initWithUUIDString:[jsonObject valueForKey:@"sid"]];
+        }
+
+        if ([[jsonObject valueForKey:@"started"] isKindOfClass:[NSString class]]) {
+            _started = [NSDate sentry_fromIso8601String:[jsonObject valueForKey:@"started"]];
+        }
+
+        if ([[jsonObject valueForKey:@"status"] isKindOfClass:[NSString class]]) {
+            NSString *status = [jsonObject valueForKey:@"status"];
+            if ([@"ok" isEqualToString:status]) {
+                _status = kSentrySessionStatusOk;
+            } else if ([@"exited" isEqualToString:status]) {
+                _status = kSentrySessionStatusExited;
+            } else if ([@"crashed" isEqualToString:status]) {
+                _status = kSentrySessionStatusCrashed;
+            } else if ([@"abnormal" isEqualToString:status]) {
+                _status = kSentrySessionStatusAbnormal;
+            }
+        }
+
+        if (nil != [jsonObject valueForKey:@"seq"]) {
+            _sequence = [[jsonObject valueForKey:@"seq"] unsignedIntegerValue];
+        }
+
+        if (nil != [jsonObject valueForKey:@"errors"]) {
+            _errors = [[jsonObject valueForKey:@"errors"] unsignedIntegerValue];
+        }
+
+        if ([[jsonObject valueForKey:@"did"] isKindOfClass:[NSString class]]) {
+            _distinctId = [jsonObject valueForKey:@"did"];
+        }
+
         id init = [jsonObject valueForKey:@"init"];
         if (nil != init) {
             _init = init;
         }
-        NSNumber *duration = [jsonObject valueForKey:@"duration"];
-        if (nil != duration) {
-            _duration = duration;
-        }
+
         id attrs = [jsonObject valueForKey:@"attrs"];
         if (nil != attrs) {
             _releaseName = [attrs valueForKey:@"release"];
             _environment = [attrs valueForKey:@"environment"];
```

Comment:
M: to be completely resilient, we'd need to check if those values exist, as we don't accept nullable release values, we'd discard this session altogether.

Reply:
I agree with that, but how would we achieve this? I can't return `nil`, because this would be a breaking change. Do you think we should put a check on `releaseName` before sending the session?

Reply:
on Java, if any `required` field is not set, we return null and do not capture.
https://github.com/getsentry/sentry-java/blob/main/sentry/src/main/java/io/sentry/SessionAdapter.java#L206-L209

this is just a defense mechanism though, ideally, we'd never cache an invalid session to the storage.